Sources/HTTP/HTTPServer.swift

```diff
@@ -28,31 +31,72 @@ public class HTTPServer {
     /// To process incoming requests
     public let handler: HTTPRequestHandler
 
-    private let server = PoCSocketSimpleServer()
+    /// Stop the server
+    public func stop() {
+        server?.close()
+    }
+
+    /// The port number the server is listening on
+    public var port: Int {
+        return options.port
+    }
+
+    /// Reference to the server for closing
+    private var server: TCPServer?
+
+    /// The number of current connections
+    public var connectionCount: Int
 
     /// Create an instance of the server. This needs to be followed with a call to `start(port:handler:)`
     public init(with newOptions: Options, requestHandler: @escaping HTTPRequestHandler) {
         options = newOptions
         handler = requestHandler
+        connectionCount = 0
     }
 
     /// Start the HTTP server on the given `port` number, using a `HTTPRequestHandler` to process incoming requests.
     public func start() throws {
-        try server.start(port: options.port, handler: handler)
+        let socket = try TCPSocket(isNonBlocking: true, shouldReuseAddress: true)
+        var eventLoops: [DispatchQueue] = []
+        for i in 1...8 {
+            eventLoops.append(.init(label: "org.swift.server.eventLoop.\(i)"))
+        }
+        let server = TCPServer(socket: socket, eventLoops: eventLoops)
+        print("Starting on http://localhost:\(options.port)/")
+        try server.start(hostname: "0.0.0.0", port: UInt16(options.port), backlog: 4096) { client in
+            self.connectionCount += 1
+            client.onClose = {
+                self.connectionCount -= 1
+            }
+            let streamingParser = StreamingParser(handler: self.handler, connectionCounter: self, keepAliveTimeout: 5.0)
+            streamingParser.parserConnector = client
+            client.start { byteBuffer in
+                let data = Data(byteBuffer)
+                let numParsed = streamingParser.readStream(data: data)
+                guard data.count == numParsed else {
+                    /// FIXME: better error
+                    fatalError()
+                }
+            }
+        }
+        self.server = server
     }
+}
 
-    /// Stop the server
-    public func stop() {
-        server.stop()
+extension TCPClient: ParserConnecting {
+    func queueSocketWrite(_ from: Data, completion: @escaping (Result) -> Void) {
+        write(from)
```

Comment:
Unless `write` is synchronous (contrary to the goal here), you cannot call `completion` until the call actually did complete. Otherwise you can't maintain back pressure and you are just overflowing the write buffer until no memory is left?

Reply:
shouldn't be too hard to add that :)

Reply:
Yeah, that probably isn't hard by itself (just queue them up and call them when the writes are done). The hard part (IMO) are the synchronization points for the closures in question as discussed on list. Hence the new `queue` argument to the handler (you can make it free threaded, but that would be pretty expensive).

Also, I'm not 100% but I think you may have the same issue on the reading side? (you need to pause reading until the bodyProcessor calls the done CB, e.g. to make this work reliably and at scale: https://github.com/ZeeZide/http-testserver/blob/master/Sources/http-testserver/main.swift#L61) Right?

Reply:
Yeah I think you're right. We should definitely be passing the DispatchQueue around w/ the request.